### PR TITLE
Disable 'Save' button when there are no changes

### LINKED ISSUE #5553: Hide 'CHANGES SAVED' message and 'Save' button if no changes have been made
**Is your feature request related to a problem? Please describe.**
When the editor loads, it shows 'CHANGES SAVED' and the 'save' button before any changes are made.  

**Describe the solution you'd like**
Instead of showing 'CHANGES SAVED', it should be blank until a change is made. After a change has been saved, this message should show. The 'Save' button should be disabled in all cases unless there are unsaved changes. 

**Describe alternatives you've considered**
none

**Additional context**

**Figure: Message 'CHANGES SAVED' and the 'save' button shown before anything was changed**

Screen Shot 2021-06-30 at 4 51 21 pm
**Figure: Mockup - Proposed UI**

### PR DESCRIPTION
**Pull request by bradystroud**
**Summary**

Closes #5553.

**Test plan**
Added a few tests

Screen Shot 2021-07-08 at 12 09 42 pm
**Figure: When no changes have been made or saved**

Screen Shot 2021-07-08 at 12 11 36 pm
**Figure: When changes have been saved** 

**Checklist**

Please add a `x` inside each checkbox:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] Code is formatted via running `yarn format`.
- [x] Tests are passing via running `yarn test`.
- [x] The status checks are successful (continuous integration). Those can be seen below.

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: packages/netlify-cms-core/src/components/Editor/EditorToolbar.js [hunk 1]
```diff
@@ -196,6 +196,10 @@ const SaveButton = styled(ToolbarButton)`
   ${buttons.lightBlue};
 `;
 
+const SaveButtonDisabled = styled(ToolbarButton)`
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/components/Editor/EditorToolbar.js:199]: How about adding:
```
  &[disabled] {
    ${buttons.disabled};
  }
```

to `SaveButton` and then passing `disabled={!hasChanged}` as prop.

TLDR: moving the condition to the CSS.

File: packages/netlify-cms-core/src/components/Editor/EditorToolbar.js [hunk 2]
```diff
@@ -583,9 +587,13 @@ export class EditorToolbar extends React.Component {
       (!hasUnpublishedChanges && !isModification && t('editor.editorToolbar.deletePublishedEntry'));
 
     return [
-      <SaveButton key="save-button" onClick={() => hasChanged && onPersist()}>
-        {isPersisting ? t('editor.editorToolbar.saving') : t('editor.editorToolbar.save')}
-      </SaveButton>,
+      hasChanged ? (
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/components/Editor/EditorToolbar.js:590]: See previous comment on `SaveButtonDisabled`

**Review by erezrokah — CHANGES_REQUESTED** [review]
Hi @bradystroud! Thanks for another UI improvement 🚀 
I've added a comment on the implementation.

Also I've noticed 2 issues:
1. The `CHANGES SAVED` never appears in non editorial workflow mode. Should it show up after publishing?
2. When transitioning between hiding `UNSAVED CHANGES` to showing it, there's a shift in the layout:
layout_shift

File: packages/netlify-cms-core/src/components/Editor/EditorToolbar.js (hunk 1 above)
Comment (bradystroud) [reply packages/netlify-cms-core/src/components/Editor/EditorToolbar.js]: I have changed this in my latest commits.
I also changed the cursor when hovering over a disabled button to the default instead of the pointer. This makes it more obvious that pressing the button won't do anything.

**Comment by bradystroud** [comment]
Hi @erezrokah,

> 1. The CHANGES SAVED never appears in non editorial workflow mode. Should it show up after publishing?

I guess it makes sense for it to show up after publishing, but ideally, I think the message should be different for non editorial workflows. e.g. CHANGES PUBLISHED but that should be a seperate issue.
For now, I have made it so after publishing, the changes saved message will show up

> 2. When transitioning between hiding UNSAVED CHANGES to showing it, there's a shift in the layout:

I have fixed this in my latest commit.

https://user-images.githubusercontent.com/38869720/125254814-2c8f5300-e33e-11eb-94ea-059592ff5bd7.mov

**Figure: No layout shifting**

**Review by erezrokah — CHANGES_REQUESTED** [review]
Hi @bradystroud, thanks for the follow up.

I would like to confirm the intended behavior before moving forward.

## Editorial workflow

1. A published entry with no linked draft - `changes saved` appears only after saving.
not-linked

2. A published entry with a linked draft - `changes saved` appears all the time including on first editor open.
linked

3. Unpublished draft - `changes saved` appears all the time including on first editor open.
unpublished

4. A new entry - `changes saved` appears only after saving.
new

## Simple workflow 
1. A published entry - `changes saved` appears all the time including on first editor open.
simple-published

2. A new entry - `changes saved` appears only after saving.
simple-new

Is that expected?
Should the behavior be - never show `changes saved` on first editor open, then once the user clicks save always show it until the next time the editor is opened (as long as they are not pending changes).

Please let me know what you think

**Comment by erezrokah** [comment]
As an additional note, how about separating disabling the `save` button into a separate PR? I think that part is much simpler and 100% done.

**Comment by bradystroud** [comment]
Hi @erezrokah,

> Is that expected?
> Should the behavior be - never show changes saved on first editor open, then once the user clicks save always show it until the next time the editor is opened (as long as they are not pending changes).

It was not expected - Instead we agree with your suggestion. 


> As an additional note, how about separating disabling the save button into a separate PR? I think that part is much simpler and 100% done.

Since we agree on the expected behaviour, I don't think that is necessary. I will make these changes today.

**Comment by bradystroud** [comment]
Hi @erezrokah,

I believe the expected behaviour is not possible as in the editorial workflows, the editor "reloads" and starts a new session when save is pressed.

Do you think this is a problem? When you press save, there is an alert anyway.

---

Alternatively, we could show 'CHANGES SAVED' if there is a saved draft. See this updated expected behaviour:

#### Editorial workflow 
- Blank on load (unless a draft exists)
- If there is a draft, show 'CHANGES SAVED'
- After making a change and pressing save, show 'CHANGES SAVED'
- For new entries, it will be blank until changes are made.

#### Simple workflow 
- Blank on load
- After making a change and pressing 'publish' it will show 'CHANGES SAVED'
- If you return to the same entry, it will be blank.
- For new entries, it will be blank until changes are made.

Let me know if any further clarification is needed.

Thanks,

-Brady

**Comment by erezrokah** [comment]
> Do you think this is a problem? When you press save, there is an alert anyway.

I'm not sure I understand the problem you're referring to nor the alert. Do you mean the publish alert? Saving doesn't show an alert.

Also to confirm - the editor reloads when you publish in simple workflow mode too, correct?

> Let me know if any further clarification is needed.

I think those makes senes - I would prefer to have a better indiction if a workflow entry is new or linked to an existing published entry, but that's another issue.

**Comment by bradystroud** [comment]
Hi @erezrokah,

I think it is working as expected, but it is failing a test.

I followed https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md#debugging-git-gateway to run the test locally, but it has succeeded.

I am using the same config.yml file, so I am not sure why I can't reproduce it.

**Comment by bradystroud** [comment]
Hi @erezrokah,

> Simple workflow
> 
> - Blank on load
> - After making a change and pressing 'publish' it will show 'CHANGES SAVED'
> - If you return to the same entry, it will be blank.
> - For new entries, it will be blank until changes are made.

I don't think this behaviour is possible for the same reasons mentioned in https://github.com/netlify/netlify-cms/pull/5595#issuecomment-882325298

There are 2 alternative behaviours for the **Simple Workflow**

**Simple workflow - alternative behaviour 1**
 
- Blank on load
- After making a change it will show 'UNSAVED CHANGES'
- After making a change and pressing 'publish' it won't show anything

**Simple workflow - alternative behaviour 2** (original behaviour - same as before this PR)
 
- Show 'CHANGES SAVED' on load
- After making a change it will show 'UNSAVED CHANGES'
- After making a change and pressing 'publish' it will show changes saved

Personally, I prefer **Alternative 2.**

Let me know what you think.

**Comment by erezrokah** [comment]
> **Simple workflow - alternative behaviour 2** (original behaviour - same as before this PR)
> 
> * Show 'CHANGES SAVED' on load
> * After making a change it will show 'UNSAVED CHANGES'
> * After making a change and pressing 'publish' it will show changes saved
> 
> Personally, I prefer **Alternative 2.**

Yes I prefer that too.

Would it make sense to keep this behavior for editorial workflow too for consistency?

Also, that test has been flaky on CI (I haven't been able to fix it yet) so if it passes locally you can ignore the failure.

**Comment by bradystroud** [comment]
> Yes I prefer that too.
>
> Would it make sense to keep this behavior for editorial workflow too for consistency?

Yes, I have made this change in my latest commit.

**Review by erezrokah — APPROVED** [review]
🚀
